File: src/dsp/bandPlan.ts
/**
 * Amateur Band Plan & Transmit Privilege Model
 * ============================================
 *
 * The data behind `canTransmit()` in catController.ts: which frequencies this station may put
 * a data-mode signal on, given the operator's regulatory region and licence class.
 *
 * Why this exists: nothing in this app used to check that the dial frequency plus the audio
 * offset landed inside an amateur allocation at all. Every mature digital-mode client refuses
 * to transmit out of band; z-30 would happily key a radio wherever the VFO happened to sit.
 *
 * Scope and limits - read this before trusting it:
 *
 *   * These are the segments in which *data* emissions are generally permitted. Some of an
 *     allocation is CW-only or phone-only and is deliberately absent here.
 *   * National regulations differ inside every IARU region, change over time, and carry
 *     conditions this table cannot express (power limits, secondary status, band-plan
 *     conventions, contest exclusions, LF/MF notches, 60 m channelisation).
 *   * The operator remains responsible for their own emissions. This gate exists to catch the
 *     obvious mistake - a mistuned VFO, a wrong band button - not to substitute for knowing
 *     your licence.
 *
 * `verifiedOn` records when each region's entries were last checked against the regulator's
 * own published table, so a stale plan is visible rather than assumed current.
 */

/** Regulatory framework the operator is licensed under. */
export type RegulatoryRegion = 'IARU_R1' | 'IARU_R2' | 'IARU_R3' | 'US';

/**
 * Licence class. `FULL` means "the full privileges of a general amateur licence in this
 * region"; the US classes model the FCC's sub-band structure, which is the case where class
 * actually changes which frequencies are legal.
 */
export type LicenseClass = 'FULL' | 'US_EXTRA' | 'US_ADVANCED' | 'US_GENERAL' | 'US_TECHNICIAN';

export interface BandSegment {
  /** Band label, matching HAM_BANDS where possible. */
  band: string;
  /** Inclusive lower edge in Hz. */
  startHz: number;
  /** Inclusive upper edge in Hz. */
  endHz: number;
  /** Licence classes for which this segment is available. */
  classes: LicenseClass[];
  /** Anything an operator should know about this segment. */
  note?: string;
}

export interface RegionBandPlan {
  region: RegulatoryRegion;
  displayName: string;
  /** ISO date on which these entries were last checked against the published allocation. */
  verifiedOn: string;
  /** Licence classes selectable within this region. */
  licenseClasses: LicenseClass[];
  segments: BandSegment[];
}

const FULL: LicenseClass[] = ['FULL'];

/** IARU Region 1: Europe, Africa, the Middle East and northern Asia. */
const REGION_1: RegionBandPlan = {
  region: 'IARU_R1',
  displayName: 'IARU Region 1 (Europe / Africa / Middle East)',
  verifiedOn: '2026-08-30',
  licenseClasses: FULL,
  segments: [
    { band: '160m', startHz: 1810000, endHz: 2000000, classes: FULL, note: 'Upper edge varies widely by country.' },
    { band: '80m', startHz: 3500000, endHz: 3800000, classes: FULL },
    { band: '60m', startHz: 5351500, endHz: 5366500, classes: FULL, note: 'Secondary, 15 W EIRP in most administrations.' },
    { band: '40m', startHz: 7000000, endHz: 7200000, classes: FULL },
    { band: '30m', startHz: 10100000, endHz: 10150000, classes: FULL, note: 'Secondary; CW and data only.' },
    { band: '20m', startHz: 14000000, endHz: 14350000, classes: FULL },
    { band: '17m', startHz: 18068000, endHz: 18168000, classes: FULL },
    { band: '15m', startHz: 21000000, endHz: 21450000, classes: FULL },
    { band: '12m', startHz: 24890000, endHz: 24990000, classes: FULL },
    { band: '10m', startHz: 28000000, endHz: 29700000, classes: FULL },
    { band: '6m', startHz: 50000000, endHz: 52000000, classes: FULL, note: 'Allocation above 52 MHz varies by country.' },
    { band: '2m', startHz: 144000000, endHz: 146000000, classes: FULL },
    { band: '70cm', startHz: 430000000, endHz: 440000000, classes: FULL },
  ],
};

/** IARU Region 2: the Americas. Use the US plan instead if licensed by the FCC. */
const REGION_2: RegionBandPlan = {
  region: 'IARU_R2',
  displayName: 'IARU Region 2 (Americas, non-US licence)',
  verifiedOn: '2026-08-30',
  licenseClasses: FULL,
  segments: [
    { band: '160m', startHz: 1800000, endHz: 2000000, classes: FULL },
    { band: '80m', startHz: 3500000, endHz: 4000000, classes: FULL },
    { band: '60m', startHz: 5351500, endHz: 5366500, classes: FULL, note: 'Secondary; channelised in some administrations.' },
    { band: '40m', startHz: 7000000, endHz: 7300000, classes: FULL },
    { band: '30m', startHz: 10100000, endHz: 10150000, classes: FULL, note: 'Secondary; CW and data only.' },
    { band: '20m', startHz: 14000000, endHz: 14350000, classes: FULL },
    { band: '17m', startHz: 18068000, endHz: 18168000, classes: FULL },
    { band: '15m', startHz: 21000000, endHz: 21450000, classes: FULL },
    { band: '12m', startHz: 24890000, endHz: 24990000, classes: FULL },
    { band: '10m', startHz: 28000000, endHz: 29700000, classes: FULL },
    { band: '6m', startHz: 50000000, endHz: 54000000, classes: FULL },
    { band: '2m', startHz: 144000000, endHz: 148000000, classes: FULL },
    { band: '70cm', startHz: 430000000, endHz: 450000000, classes: FULL },
  ],
};

/** IARU Region 3: Asia-Pacific. */
const REGION_3: RegionBandPlan = {
  region: 'IARU_R3',
  displayName: 'IARU Region 3 (Asia / Pacific)',
  verifiedOn: '2026-08-30',
  licenseClasses: FULL,
  segments: [
    { band: '160m', startHz: 1800000, endHz: 2000000, classes: FULL, note: 'Sub-band allocation varies widely.' },
    { band: '80m', startHz: 3500000, endHz: 3900000, classes: FULL },
    { band: '60m', startHz: 5351500, endHz: 5366500, classes: FULL, note: 'Not available in every administration.' },
    { band: '40m', startHz: 7000000, endHz: 7200000, classes: FULL },
    { band: '30m', startHz: 10100000, endHz: 10150000, classes: FULL, note: 'Secondary; CW and data only.' },
    { band: '20m', startHz: 14000000, endHz: 14350000, classes: FULL },
    { band: '17m', startHz: 18068000, endHz: 18168000, classes: FULL },
    { band: '15m', startHz: 21000000, endHz: 21450000, classes: FULL },
    { band: '12m', startHz: 24890000, endHz: 24990000, classes: FULL },
    { band: '10m', startHz: 28000000, endHz: 29700000, classes: FULL },
    { band: '6m', startHz: 50000000, endHz: 54000000, classes: FULL },
    { band: '2m', startHz: 144000000, endHz: 148000000, classes: FULL },
    { band: '70cm', startHz: 430000000, endHz: 440000000, classes: FULL },
  ],
};

const US_ALL: LicenseClass[] = ['US_EXTRA', 'US_ADVANCED', 'US_GENERAL', 'US_TECHNICIAN'];
const US_HF: LicenseClass[] = ['US_EXTRA', 'US_ADVANCED', 'US_GENERAL'];

/**
 * United States, FCC Part 97 RTTY/data sub-bands (97.301 / 97.305). Technicians hold data
 * privileges on 10 m and everything at 6 m and above; their HF privileges below 10 m are CW
 * only, so they do not appear in those segments.
 */
const US: RegionBandPlan = {
  region: 'US',
  displayName: 'United States (FCC Part 97)',
  verifiedOn: '2026-08-30',
  licenseClasses: US_ALL,
  segments: [
    { band: '160m', startHz: 1800000, endHz: 2000000, classes: US_HF },
    { band: '80m', startHz: 3500000, endHz: 3600000, classes: ['US_EXTRA'] },
    { band: '80m', startHz: 3525000, endHz: 3600000, classes: ['US_ADVANCED', 'US_GENERAL'] },
    { band: '60m', startHz: 5332000, endHz: 5405000, classes: US_HF, note: 'Channelised: 5332.0, 5348.0, 5358.5, 5373.0 and 5405.0 kHz centres only, 100 W ERP.' },
    { band: '40m', startHz: 7000000, endHz: 7125000, classes: ['US_EXTRA'] },
    { band: '40m', startHz: 7025000, endHz: 7125000, classes: ['US_ADVANCED', 'US_GENERAL'] },
    { band: '30m', startHz: 10100000, endHz: 10150000, classes: US_HF, note: 'Secondary; 200 W PEP maximum.' },
    { band: '20m', startHz: 14000000, endHz: 14150000, classes: ['US_EXTRA'] },
    { band: '20m', startHz: 14025000, endHz: 14150000, classes: ['US_ADVANCED', 'US_GENERAL'] },
    { band: '17m', startHz: 18068000, endHz: 18110000, classes: US_HF },
    { band: '15m', startHz: 21000000, endHz: 21200000, classes: ['US_EXTRA'] },
    { band: '15m', startHz: 21025000, endHz: 21200000, classes: ['US_ADVANCED', 'US_GENERAL'] },
    { band: '12m', startHz: 24890000, endHz: 24930000, classes: US_HF },
    { band: '10m', startHz: 28000000, endHz: 28300000, classes: US_ALL, note: 'Technicians: 200 W PEP maximum.' },
    { band: '6m', startHz: 50000000, endHz: 54000000, classes: US_ALL },
    { band: '2m', startHz: 144000000, endHz: 148000000, classes: US_ALL },
    { band: '70cm', startHz: 420000000, endHz: 450000000, classes: US_ALL, note: 'Secondary; geographic restrictions apply near the Canadian border and certain radar sites.' },
  ],
};

export const BAND_PLANS: Record<RegulatoryRegion, RegionBandPlan> = {
  IARU_R1: REGION_1,
  IARU_R2: REGION_2,
  IARU_R3: REGION_3,
  US,
};

export const REGULATORY_REGION_OPTIONS: { id: RegulatoryRegion; label: string }[] = [
  { id: 'US', label: BAND_PLANS.US.displayName },
  { id: 'IARU_R1', label: BAND_PLANS.IARU_R1.displayName },
  { id: 'IARU_R2', label: BAND_PLANS.IARU_R2.displayName },
  { id: 'IARU_R3', label: BAND_PLANS.IARU_R3.displayName },
];

export const LICENSE_CLASS_LABELS: Record<LicenseClass, string> = {
  FULL: 'Full amateur privileges',
  US_EXTRA: 'Amateur Extra',
  US_ADVANCED: 'Advanced',
  US_GENERAL: 'General',
  US_TECHNICIAN: 'Technician',
};

/**
 * Finds the segment that fully contains an emission of `bandwidthHz` centred on
 * `frequencyHz`, for the given region and class. Returns null when any part of the emission
 * falls outside every permitted segment.
 *
 * The check is against the emission's EDGES, not its centre. A z-30 signal is 50 Hz wide, so a
 * station whose centre sits 10 Hz inside a band edge is radiating more than half its power
 * outside the band - legally out of band, and exactly the mistake an operator tuning up to the
 * edge of a segment makes. Passing bandwidthHz = 0 recovers the old centre-only behaviour and
 * is only appropriate for a carrier of negligible width.
 */
export function findPermittedSegment(
  region: RegulatoryRegion,
  licenseClass: LicenseClass,
  frequencyHz: number,
  bandwidthHz: number = 0
): BandSegment | null {
  const plan = BAND_PLANS[region];
  if (!plan) return null;
  const half = Math.max(0, bandwidthHz) / 2;
  const lowEdgeHz = frequencyHz - half;
  const highEdgeHz = frequencyHz + half;
  for (const segment of plan.segments) {
    if (
      lowEdgeHz >= segment.startHz &&
      highEdgeHz <= segment.endHz &&
      segment.classes.includes(licenseClass)
    ) {
      return segment;
    }
  }
  return null;
}

/**
 * The permitted segment nearest to `frequencyHz`, with the distance to it, so a refusal can
 * tell the operator how far out of band they are instead of only that they are.
 */
export function nearestPermittedSegment(
  region: RegulatoryRegion,
  licenseClass: LicenseClass,
  frequencyHz: number
): { segment: BandSegment; distanceHz: number } | null {
  const plan = BAND_PLANS[region];
  if (!plan) return null;
  let best: { segment: BandSegment; distanceHz: number } | null = null;
  for (const segment of plan.segments) {
    if (!segment.classes.includes(licenseClass)) continue;
    const distanceHz =
      frequencyHz < segment.startHz
        ? segment.startHz - frequencyHz
        : frequencyHz > segment.endHz
          ? frequencyHz - segment.endHz
          : 0;
    if (!best || distanceHz < best.distanceHz) best = { segment, distanceHz };
  }
  return best;
}

/**
 * Syntactic validation of an amateur callsign.
 *
 * Deliberately structural rather than exhaustive: one to two prefix characters (at least one a
 * letter), a digit, and a one-to-four-character suffix, with optional `/P`-style appendages.
 * It rejects an empty box, a name typed into the callsign field, and obvious typos - which is
 * what the transmit gate needs. It cannot tell whether a well-formed callsign was actually
 * issued to the operator.
 */
export function isValidCallsign(callsign: string): boolean {
  const trimmed = (callsign || '').trim().toUpperCase();
  if (!trimmed) return false;
  return /^(?:[A-Z0-9]{1,3}\/)?[A-Z0-9]{1,3}\d[A-Z]{1,4}(?:\/[A-Z0-9]{1,4})?$/.test(trimmed) &&
    /[A-Z]/.test(trimmed.split('/')[0] || trimmed);
}

File: crates/z30-engine/src/bandplan.rs
//! Amateur band plan and transmit privileges: which frequencies this station may put a data
//! emission on, given its regulatory region and licence class. A port of `src/dsp/bandPlan.ts`
//! (the data behind the transmit gate), with the same scope limits: data segments only,
//! national variations not modelled, and the operator remains responsible for their emissions.

use serde::{Deserialize, Serialize};

/// Regulatory framework the operator is licensed under.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum Region {
    /// IARU Region 1: Europe, Africa, Middle East, northern Asia.
    IaruR1,
    /// IARU Region 2: the Americas (non-US licence).
    IaruR2,
    /// IARU Region 3: Asia-Pacific.
    IaruR3,
    /// United States, FCC Part 97.
    Us,
}

/// Licence class.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum LicenseClass {
    /// Full privileges of a general amateur licence in an IARU region.
    Full,
    /// US Amateur Extra.
    UsExtra,
    /// US Advanced.
    UsAdvanced,
    /// US General.
    UsGeneral,
    /// US Technician.
    UsTechnician,
}

impl Region {
    /// Parses the TypeScript identifiers (`IARU_R1`, `US`, ...).
    pub fn from_legacy(s: &str) -> Option<Region> {
        Some(match s {
            "IARU_R1" => Region::IaruR1,
            "IARU_R2" => Region::IaruR2,
            "IARU_R3" => Region::IaruR3,
            "US" => Region::Us,
            _ => return None,
        })
    }

    /// Display name.
    pub fn display_name(self) -> &'static str {
        match self {
            Region::IaruR1 => "IARU Region 1 (Europe / Africa / Middle East)",
            Region::IaruR2 => "IARU Region 2 (Americas, non-US licence)",
            Region::IaruR3 => "IARU Region 3 (Asia / Pacific)",
            Region::Us => "United States (FCC Part 97)",
        }
    }

    /// Licence classes that apply in this region.
    pub fn license_classes(self) -> &'static [LicenseClass] {
        match self {
            Region::Us => &[LicenseClass::UsExtra, LicenseClass::UsAdvanced, LicenseClass::UsGeneral, LicenseClass::UsTechnician],
            _ => &[LicenseClass::Full],
        }
    }

    /// When the region's table was last checked against the regulator's published allocation.
    pub fn verified_on(self) -> &'static str {
        "2026-08-30"
    }
}

impl LicenseClass {
    /// Parses the TypeScript identifiers (`FULL`, `US_EXTRA`, ...).
    pub fn from_legacy(s: &str) -> Option<LicenseClass> {
        Some(match s {
            "FULL" => LicenseClass::Full,
            "US_EXTRA" => LicenseClass::UsExtra,
            "US_ADVANCED" => LicenseClass::UsAdvanced,
            "US_GENERAL" => LicenseClass::UsGeneral,
            "US_TECHNICIAN" => LicenseClass::UsTechnician,
            _ => return None,
        })
    }
}

/// A data-mode segment.
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct Segment {
    /// Band label.
    pub band: &'static str,
    /// Inclusive lower edge, Hz.
    pub start_hz: u64,
    /// Inclusive upper edge, Hz.
    pub end_hz: u64,
    /// Classes that may use it.
    pub classes: &'static [LicenseClass],
}

use LicenseClass::*;
const FULL: &[LicenseClass] = &[Full];
const US_ALL: &[LicenseClass] = &[UsExtra, UsAdvanced, UsGeneral, UsTechnician];
const US_HF: &[LicenseClass] = &[UsExtra, UsAdvanced, UsGeneral];
const US_EX: &[LicenseClass] = &[UsExtra];
const US_AG: &[LicenseClass] = &[UsAdvanced, UsGeneral];

macro_rules! seg {
    ($b:expr, $s:expr, $e:expr, $c:expr) => {
        Segment { band: $b, start_hz: $s, end_hz: $e, classes: $c }
    };
}

const R1: &[Segment] = &[
    seg!("160m", 1_810_000, 2_000_000, FULL),
    seg!("80m", 3_500_000, 3_800_000, FULL),
    seg!("60m", 5_351_500, 5_366_500, FULL),
    seg!("40m", 7_000_000, 7_200_000, FULL),
    seg!("30m", 10_100_000, 10_150_000, FULL),
    seg!("20m", 14_000_000, 14_350_000, FULL),
    seg!("17m", 18_068_000, 18_168_000, FULL),
    seg!("15m", 21_000_000, 21_450_000, FULL),
    seg!("12m", 24_890_000, 24_990_000, FULL),
    seg!("10m", 28_000_000, 29_700_000, FULL),
    seg!("6m", 50_000_000, 52_000_000, FULL),
    seg!("2m", 144_000_000, 146_000_000, FULL),
    seg!("70cm", 430_000_000, 440_000_000, FULL),
];

const R2: &[Segment] = &[
    seg!("160m", 1_800_000, 2_000_000, FULL),
    seg!("80m", 3_500_000, 4_000_000, FULL),
    seg!("60m", 5_351_500, 5_366_500, FULL),
    seg!("40m", 7_000_000, 7_300_000, FULL),
    seg!("30m", 10_100_000, 10_150_000, FULL),
    seg!("20m", 14_000_000, 14_350_000, FULL),
    seg!("17m", 18_068_000, 18_168_000, FULL),
    seg!("15m", 21_000_000, 21_450_000, FULL),
    seg!("12m", 24_890_000, 24_990_000, FULL),
    seg!("10m", 28_000_000, 29_700_000, FULL),
    seg!("6m", 50_000_000, 54_000_000, FULL),
    seg!("2m", 144_000_000, 148_000_000, FULL),
    seg!("70cm", 430_000_000, 450_000_000, FULL),
];

const R3: &[Segment] = &[
    seg!("160m", 1_800_000, 2_000_000, FULL),
    seg!("80m", 3_500_000, 3_900_000, FULL),
    seg!("60m", 5_351_500, 5_366_500, FULL),
    seg!("40m", 7_000_000, 7_200_000, FULL),
    seg!("30m", 10_100_000, 10_150_000, FULL),
    seg!("20m", 14_000_000, 14_350_000, FULL),
    seg!("17m", 18_068_000, 18_168_000, FULL),
    seg!("15m", 21_000_000, 21_450_000, FULL),
    seg!("12m", 24_890_000, 24_990_000, FULL),
    seg!("10m", 28_000_000, 29_700_000, FULL),
    seg!("6m", 50_000_000, 54_000_000, FULL),
    seg!("2m", 144_000_000, 148_000_000, FULL),
    seg!("70cm", 430_000_000, 440_000_000, FULL),
];

const US: &[Segment] = &[
    seg!("160m", 1_800_000, 2_000_000, US_HF),
    seg!("80m", 3_500_000, 3_600_000, US_EX),
    seg!("80m", 3_525_000, 3_600_000, US_AG),
    seg!("60m", 5_332_000, 5_405_000, US_HF),
    seg!("40m", 7_000_000, 7_125_000, US_EX),
    seg!("40m", 7_025_000, 7_125_000, US_AG),
    seg!("30m", 10_100_000, 10_150_000, US_HF),
    seg!("20m", 14_000_000, 14_150_000, US_EX),
    seg!("20m", 14_025_000, 14_150_000, US_AG),
    seg!("17m", 18_068_000, 18_110_000, US_HF),
    seg!("15m", 21_000_000, 21_200_000, US_EX),
    seg!("15m", 21_025_000, 21_200_000, US_AG),
    seg!("12m", 24_890_000, 24_930_000, US_HF),
    seg!("10m", 28_000_000, 28_300_000, US_ALL),
    seg!("6m", 50_000_000, 54_000_000, US_ALL),
    seg!("2m", 144_000_000, 148_000_000, US_ALL),
    seg!("70cm", 420_000_000, 450_000_000, US_ALL),
];

/// The data segments of a region.
pub fn segments(region: Region) -> &'static [Segment] {
    match region {
        Region::IaruR1 => R1,
        Region::IaruR2 => R2,
        Region::IaruR3 => R3,
        Region::Us => US,
    }
}

/// The segment that contains the whole emission `[f - bw/2, f + bw/2]`, if any. Checked against
/// the emission's edges, not its centre: a 50 Hz signal centred 10 Hz inside a band edge is
/// radiating out of band.
pub fn find_permitted_segment(region: Region, class: LicenseClass, freq_hz: f64, bandwidth_hz: f64) -> Option<Segment> {
    let half = bandwidth_hz.max(0.0) / 2.0;
    let (lo, hi) = (freq_hz - half, freq_hz + half);
    segments(region).iter().copied().find(|s| lo >= s.start_hz as f64 && hi <= s.end_hz as f64 && s.classes.contains(&class))
}

/// The nearest permitted segment and the distance to it, so a refusal can say how far out the
/// station is.
pub fn nearest_permitted_segment(region: Region, class: LicenseClass, freq_hz: f64) -> Option<(Segment, f64)> {
    segments(region)
        .iter()
        .filter(|s| s.classes.contains(&class))
        .map(|s| {
            let d = if freq_hz < s.start_hz as f64 {
                s.start_hz as f64 - freq_hz
            } else if freq_hz > s.end_hz as f64 {
                freq_hz - s.end_hz as f64
            } else {
                0.0
            };
            (*s, d)
        })
        .min_by(|a, b| a.1.partial_cmp(&b.1).unwrap())
}

/// Structural callsign validation, the rule `isValidCallsign` in bandPlan.ts and
/// `station_settings.py` share (`tests/vectors/callsign_vectors.json`):
/// `^(?:[A-Z0-9]{1,3}/)?[A-Z0-9]{1,3}\d[A-Z]{1,4}(?:/[A-Z0-9]{1,4})?$` with a letter in the
/// first `/`-separated part. Structural only: whether v1 can carry it is a separate question
/// (`z30_protocol::codec::Callsign::new`), and the transmit gate asks both.
pub fn is_valid_callsign(call: &str) -> bool {
    let t = call.trim().to_ascii_uppercase();
    if t.is_empty() {
        return false;
    }
    let first_part = t.split('/').next().unwrap_or(&t);
    if !first_part.bytes().any(|b| b.is_ascii_uppercase()) {
        return false;
    }
    let alnum =
        |s: &str, lo: usize, hi: usize| (lo..=hi).contains(&s.len()) && s.bytes().all(|b| b.is_ascii_uppercase() || b.is_ascii_digit());
    let parts: Vec<&str> = t.split('/').collect();
    let (prefix, core, suffix) = match parts.len() {
        1 => (None, parts[0], None),
        2 => {
            // Either PREFIX/CORE or CORE/SUFFIX; the regex accepts whichever parse matches.
            if core_ok(parts[1]) && alnum(parts[0], 1, 3) {
                (Some(parts[0]), parts[1], None)
            } else {
                (None, parts[0], Some(parts[1]))
            }
        }
        3 => (Some(parts[0]), parts[1], Some(parts[2])),
        _ => return false,
    };
    prefix.is_none_or(|p| alnum(p, 1, 3)) && core_ok(core) && suffix.is_none_or(|s| alnum(s, 1, 4))
}

/// `[A-Z0-9]{1,3}\d[A-Z]{1,4}` - tried at every split, as a backtracking regex would.
fn core_ok(core: &str) -> bool {
    let b = core.as_bytes();
    let suffix_len = b.iter().rev().take_while(|c| c.is_ascii_uppercase()).count();
    for sl in 1..=suffix_len.min(4) {
        if b.len() < sl + 2 {
            continue;
        }
        let digit_idx = b.len() - sl - 1;
        if !b[digit_idx].is_ascii_digit() {
            continue;
        }
        let pre = &b[..digit_idx];
        if (1..=3).contains(&pre.len()) && pre.iter().all(|c| c.is_ascii_uppercase() || c.is_ascii_digit()) {
            return true;
        }
    }
    false
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn edges_not_centres() {
        // 14.074 MHz dial + 1500 Hz audio is well inside 20 m.
        assert!(find_permitted_segment(Region::IaruR1, LicenseClass::Full, 14_075_500.0, 50.0).is_some());
        // Centre 10 Hz inside the upper edge of 20 m in R1: half the emission is outside.
        assert!(find_permitted_segment(Region::IaruR1, LicenseClass::Full, 14_349_990.0, 50.0).is_none());
        // US General: 14.000-14.025 is Extra-only.
        assert!(find_permitted_segment(Region::Us, LicenseClass::UsGeneral, 14_010_000.0, 50.0).is_none());
        assert!(find_permitted_segment(Region::Us, LicenseClass::UsExtra, 14_010_000.0, 50.0).is_some());
        // US Technician has no 20 m data privileges.
        assert!(find_permitted_segment(Region::Us, LicenseClass::UsTechnician, 14_076_000.0, 50.0).is_none());
    }

    #[test]
    fn nearest_segment_reports_distance() {
        let (s, d) = nearest_permitted_segment(Region::IaruR1, LicenseClass::Full, 14_400_000.0).unwrap();
        assert_eq!(s.band, "20m");
        assert_eq!(d, 50_000.0);
    }
}
